File: examples/python/level-2-streaming.py
from lomond import WebSocket
from lomond.persist import persist

def parse_snapshot(message):
    sub_sections = message.split(':');
    message_key = sub_sections[0];
    bids_str = sub_sections[1];
    asks_str = sub_sections[2];
    full_OB = {}
    full_OB["bids"] = [[float(item) for item in position.split('~')] for position in bids_str.split(',')]
    full_OB["asks"] = [[float(item) for item in position.split('~')]  for position in asks_str.split(',')]
    return full_OB;


def get_change_flag(flag): 
    if (flag == '1'):
        return 'add'
    elif (flag == '2'):
        return 'remove'
    elif (flag == '4'):
        return 'update'
    return 'unknown'; #This should not happen

def parse_level_2(message):
    data_array = message.split('~');
    data_obj = {
        "message_type": data_array[0],
        "exchange": data_array[1],
        "fromsymbol": data_array[2],
        "tosymbol": data_array[3],
        "side": 'bid' if data_array[4] == '1' else 'ask',
        "change": get_change_flag(data_array[5]),
        "sequence": int(data_array[6]),
        "price": float(data_array[7]),
        "quantity": float(data_array[8])
    }
    return data_obj

cryptocompare_ws = WebSocket('wss://streaming.cryptocompare.com')
subs = {
    "action": "SubAdd",
    "subs": ["8~bitfinex~BTC~USD"],
    "api_key": "YOUR-API-KEY",
    "format": "streamer"
}

for event in persist(cryptocompare_ws):
    if event.name == "ready":
        print('ready')
        cryptocompare_ws.send_json(subs)
    if event.name == "text":
        print(event.text)
        messages = event.text.split("|")
        messages = messages[:-1] #Last element is empty string
        for message in messages:
            if message == "999~HEARTBEAT|":
                print(message)
            elif message.split("~")[0] == "9":
                print(parse_snapshot(message))
                continue
            else:
                print(parse_level_2(message))

            

File: examples/nodejs/level-2-streaming.js
const ws = require('ws');


function parseSnapshot(message) {
    let subSections = message.split(':');
    let messageKey = subSections[0];
    let bidsString = subSections[1];
    let asksString = subSections[2];
    
    let fullOB = {}
    fullOB.bids = bidsString.split(',').map(function(item){
        return item.split('~').map(parseFloat);
    });
    fullOB.asks = asksString.split(',').map(function(item){
        return item.split('~').map(parseFloat);
    });
    return fullOB;
}

function getChangeFlag(flag) {
    if (flag === '1') {
        return 'add';
    } else if (flag === '2') {
        return 'remove';
    } else if (flag === '4') {
        return 'update';
    }
    return 'unknown'; //This should not happen
}

function parseLevel2(message) {
    let dataArray = message.split('~');
    let dataObj = {
        type: dataArray[0],
        exchange: dataArray[1],
        fromsymbol: dataArray[2],
        tosymbol: dataArray[3],
        side: dataArray[4] === '1' ? 'bid' : 'ask',
        change: getChangeFlag(dataArray[5]),
        sequence: parseInt(dataArray[6]),
        price: parseFloat(dataArray[7]),
        quantity: parseFloat(dataArray[8])
    }
    return dataObj;
}

let cryptocompareWS = new ws('wss://streaming.cryptocompare.com')

let subs = {
    "action": "SubAdd",
    "subs": ["8~bitfinex~BTC~USD"],
    "api_key": "YOUR-API-KEY",
    "format": "streamer"
}

cryptocompareWS.on('open', function(){
    console.log('connection established')
    cryptocompareWS.send(JSON.stringify(subs));
});

cryptocompareWS.on('error', function(error){
    console.log(error)
});

cryptocompareWS.on('close', function(){
    console.log('disconnected');
});

cryptocompareWS.on('message', function(data){
    console.log(data)
    if (data === '999~HEARTBEAT|') {
        console.log(data);
        return;
    }
    let messages = data.split('|');
    messages.pop(); //Remove last element as it is always an empty string
    for (let message of messages) {
        if (message.split('~')[0] === '9') {
            let snapshot = parseSnapshot(message);
            console.log(snapshot)
            continue;
        }
        let parsedUpdate = parseLevel2(message);
        
        console.log(JSON.stringify(parsedUpdate))
    }
});

